db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -4014,6 +4014,11 @@ Status DBImpl::WaitForCompact(
          unscheduled_flushes_) &&
         (error_handler_.GetBGError().ok())) {
       bg_cv_.Wait();
+    } else if (wait_for_compact_options.close_db) {
+      mutex_.Unlock();
+      Status s = Close();
```

Comment:
From my internal description: "I'm not sure that WaitForCompact() followed by Close(), by the only user thread interacting with the DB, is sufficient to guarantee no background work touching storage is cancelled on close (e.g. periodic stats)."

By releasing the DB mutex here we open up an opportunity (however short) for the DB to start more background work. Actually Close() also releases the db mutex in some of its work, so I think we need to CancelAllBackgroundWork() or something similar before releasing the mutex. We might need to refactor CancelAllBackgroundWork(). Hmm. Feel free to discuss possible solutions with me outside of GitHub.

Reply:
It looks like Close() will cancel all background work (including calling `CancelAllBackgroundWork()`). Then it would suffice to just call Close() here?

Reply:
The expectation from the user is that once they close the db by calling `WaitForCompact()` with `close_db` and `flush` option, they should expect no flush/compaction debt upon reopening the db.

If there's new background work added in the short period of mutex released and got canceled during `Close()`, what would happen upon re-opening the db?